gcp_variant_transforms/libs/variant_partition.py

```diff
@@ -34,59 +44,266 @@
+    self._interval_tree.addi(start, end, partition_index)
+
+  def get_partition_index(self, pos=0):
+    """Finds a region that pos falls in it, if none PARTITION_NOT_FOUND."""
+    matched_regions = self._interval_tree.search(pos)
+    # Ensure at most one region is matching to the give position.
+    assert len(matched_regions) <= 1
+    if len(matched_regions) == 1:
+      return next(iter(matched_regions)).data
+    else:
+      return PARTITION_NOT_FOUND
+
+class VariantPartition(object):
+  """Partition variants based on their reference_name and position.
+
+  This class has 2 operating modes:
+    1) No config file is given (config_file_path == None):
+       Automatically partition variants based on their reference_name.
+    2) A config file is given:
+       partitions variants based on input config file.
+  """
+
+  def __init__(self, config_file_path=None):
     if _DEFAULT_NUM_PARTITIONS <= _RESERVED_AUTO_PARTITIONS:
       raise ValueError(
           '_DEFAULT_NUM_PARTITIONS must be > _RESERVED_AUTO_PARTITIONS')
-    self._reference_name_to_partition_map = {}
+    self._num_partitions = _DEFAULT_NUM_PARTITIONS
+    # This variable determines the operation mode auto (default mode) vs config.
+    self._config_file_path_given = False
+    # Residual partition will contain all remaining variants that do not match
+    # to any other partition. -1 means it does not exist.
+    self._residual_partition_index = -1
+    self._should_keep_residual_partition = False
+    self._ref_name_to_partitions_map = defaultdict(_ChromosomePartitioner)
+    self._partition_names = {}
+
+    if config_file_path:
+      self._config_file_path_given = True
+      self._parse_config(config_file_path)
+
+  def _validate_config(self, config_file_path):
+    # type: (str) -> None
+    with FileSystems.open(config_file_path, 'r') as f:
+      try:
+        partition_configs = yaml.load(f)
+      except yaml.YAMLError as e:
+        raise ValueError('Invalid yaml file: %s' % str(e))
+    if len(partition_configs) > _MAX_NUM_PARTITIONS:
+      raise ValueError(
+          'There can be at most {} partitions but given config file '
+          'contains {}'.format(_MAX_NUM_PARTITIONS, len(partition_configs)))
+    if not partition_configs:
+      raise ValueError('There must be at least one partition in config file.')
+
+    existing_partition_names = set()
+    for partition_config in partition_configs:
+      partition = partition_config.get('partition', None)
+      if partition is None:
+        raise ValueError('Wrong yaml file format, partition field missing.')
+      regions = partition.get('regions', None)
+      if regions is None:
+        raise ValueError('Each partition must have at least one region.')
+      if len(regions) > _MAX_NUM_REGIONS:
+        raise ValueError('At most {} regions per partition, thie partition '
+                         'contains {}'.format(_MAX_NUM_REGIONS, len(regions)))
+      # Check the partition_name is unique.
```

Comment:
nit: this comment should be below this if? Again, I don't think it adds any value as the if statement is pretty self-explanatory IMO.